File: bindings/python/cpp/src/types/structured/regular/domain_descriptor.cpp
/*
 * GridTools
 *
 * Copyright (c) 2014-2021, ETH Zurich
 * All rights reserved.
 *
 * Please, refer to the LICENSE file in the root directory.
 * SPDX-License-Identifier: BSD-3-Clause
 *
 */
#include "boost/mp11/algorithm.hpp"

#include "ghex/bindings/python/types/structured/regular/domain_descriptor.hpp"

namespace gridtools {
namespace ghex {
namespace bindings {
namespace python {
namespace types {
namespace structured {
namespace regular {

void export_domain_descriptor (py::module_& m) {
    using dim_type_list = boost::mp11::mp_unique<
        boost::mp11::mp_transform<
            gridtools::ghex::bindings::python::dim_type,
            typename gridtools::ghex::bindings::python::type_list::layout_map_type>>;
    using exporter_type_list = boost::mp11::mp_transform<domain_descriptor_exporter, dim_type_list>;
    boost::mp11::mp_for_each<exporter_type_list>([&](auto exporter){ exporter(m); });
}

}
}
}
}
}
}
}

File: bindings/python/src/ghex/structured/regular/domain_descriptor.py
# -*- coding: utf-8 -*-
#
# GridTools
#
# Copyright (c) 2014-2021, ETH Zurich
# All rights reserved.
#
# Please, refer to the LICENSE file in the root directory.
# SPDX-License-Identifier: BSD-3-Clause
from __future__ import annotations
from typing import TYPE_CHECKING

from ghex.utils.cpp_wrapper_utils import CppWrapper

if TYPE_CHECKING:
    from ghex.utils.index_space import ProductSet


class DomainDescriptor(CppWrapper):
    def __init__(self, id_: int, sub_domain_indices: ProductSet):

        super(DomainDescriptor, self).__init__(
            (
                "gridtools::ghex::structured::regular::domain_descriptor",
                "int",
                f"std::integral_constant<int, {sub_domain_indices.dim}> ",
            ),
            id_,
            sub_domain_indices[0, 0, 0],
            sub_domain_indices[-1, -1, -1],
        )
